rbx_types/src/attributes/mod.rs

```diff
@@ -0,0 +1,141 @@
+use crate::variant::VariantType;
+use std::convert::TryFrom;
+use thiserror::Error;
+
+mod reader;
+mod writer;
+
+pub use reader::get_attributes;
+pub use writer::attributes_from_map;
+
+macro_rules! create_attribute_type {
+    ({
+        $(
+            $key:ident = $number:tt,
+        )+
+    }) => {
+        #[derive(Clone, Copy, Debug, Eq, PartialEq)]
+        pub(crate) enum AttributeType {
+            $(
+                $key = $number,
+            )+
+        }
+
+        impl TryFrom<VariantType> for AttributeType {
+            type Error = AttributeError;
+
+            fn try_from(variant_type: VariantType) -> Result<Self, Self::Error> {
+                match variant_type {
+                    $(
+                        VariantType::$key => Ok(AttributeType::$key),
+                    )+
+
+                    _ => Err(AttributeError::InvalidVariantType),
+                }
+            }
+        }
+
+        impl TryFrom<u8> for AttributeType {
+            type Error = AttributeError;
+
+            fn try_from(byte: u8) -> Result<Self, Self::Error> {
+                match byte {
+                    $(
+                        $number => Ok(Self::$key),
+                    )+
+
+                    other => Err(AttributeError::InvalidValueType(other))
+                }
+            }
+        }
+    };
+}
+
+create_attribute_type!({
+    BinaryString = 0x02,
+    Bool = 0x03,
+    Float32 = 0x05,
+    Float64 = 0x06,
+    UDim = 0x09,
+    UDim2 = 0x0A,
+    BrickColor = 0x0E,
+    Color3 = 0x0F,
+    Vector2 = 0x10,
+    Vector3 = 0x11,
+    NumberSequence = 0x17,
+    ColorSequence = 0x19,
+    NumberRange = 0x1B,
+    Rect = 0x1C,
+});
+
+#[derive(Debug, Clone, Error, PartialEq, Eq)]
+pub enum AttributeError {
```

Comment:
I think that public error enums are usually a trap as it hurts our ability to refactor. Usually, consumers don't need to match on these. We should make this type `pub(crate)` at most and expose a top-level error type that's just an opaque error struct.

`thiserror` supports this pattern pretty well.